interfaces/prompting/constraints.go

```diff
@@ -30,62 +31,93 @@ import (
+		return false, prompting_errors.NewInvalidPathPatternError(c.PathPattern.String(), err.Error())
 	}
-	return c.validatePermissions(iface)
+	return match, nil
 }
 
-// validatePermissions checks that the permissions for the given constraints
-// are valid for the given interface. If not, returns an error, otherwise
-// ensures that the permissions are in the order in which they occur in the
-// list of available permissions for that interface.
-func (c *Constraints) validatePermissions(iface string) error {
-	availablePerms, ok := interfacePermissionsAvailable[iface]
-	if !ok {
-		return prompting_errors.NewInvalidInterfaceError(iface, availableInterfaces())
-	}
-	permsSet := make(map[string]bool, len(c.Permissions))
-	var invalidPerms []string
-	for _, perm := range c.Permissions {
-		if !strutil.ListContains(availablePerms, perm) {
-			invalidPerms = append(invalidPerms, perm)
-			continue
+// ContainPermissions returns true if the permission map in the constraints
+// includes every one of the given permissions.
+//
+// This method is only intended to be called on constraints which have just
+// been created from a reply, to check that the reply covers the request.
+func (c *Constraints) ContainPermissions(permissions []string) bool {
+	for _, perm := range permissions {
+		if _, exists := c.Permissions[perm]; !exists {
+			return false
 		}
-		permsSet[perm] = true
 	}
-	if len(invalidPerms) > 0 {
-		return prompting_errors.NewInvalidPermissionsError(iface, invalidPerms, availablePerms)
+	return true
+}
+
+// ToRuleConstraints validates the receiving Constraints and converts it to
+// RuleConstraints. If the constraints are not valid with respect to the given
+// interface, returns an error.
+func (c *Constraints) ToRuleConstraints(iface string, currTime time.Time) (*RuleConstraints, error) {
+	if c.PathPattern == nil {
+		return nil, prompting_errors.NewInvalidPathPatternError("", "no path pattern")
 	}
-	if len(permsSet) == 0 {
-		return prompting_errors.NewPermissionsListEmptyError(iface, availablePerms)
+	rulePermissions, err := c.Permissions.toRulePermissionMap(iface, currTime)
+	if err != nil {
+		return nil, err
 	}
-	newPermissions := make([]string, 0, len(permsSet))
-	for _, perm := range availablePerms {
-		if exists := permsSet[perm]; exists {
-			newPermissions = append(newPermissions, perm)
-		}
+	ruleConstraints := &RuleConstraints{
+		PathPattern: c.PathPattern,
+		Permissions: rulePermissions,
 	}
-	c.Permissions = newPermissions
-	return nil
+	return ruleConstraints, nil
+}
+
+// RuleConstraints hold information about the applicability of an existing rule
+// to particular paths or permissions. A request will be matched by the rule
+// constraints if the requested path is matched by the path pattern (according
+// to bash's globstar matching) and one or more requested permissions are denied
+// in the permission map, or all of the requested permissions are allowed in the
+// map.
```

Comment:
not sure it's useful to repeat the same thing in every doc comment.
```suggestion
// to particular paths or permissions.
```

Reply:
Agreed, though I think I'd rather have the explanation here, as `RuleConstraints` are what is used to actually match against prompts. `Constraints` are the more ergonomic (using `Duration` instead of `Expiration`) form of `RuleConstraints`, and they're converted to the latter before being matched against any requests. I'll remove the verbose explanation from `Constraints` and note that `Constraints` are converted into `RuleConstraints`.